File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class Main {
	static int N;
	static int[][] map;
	static int[] dr = { -1, 0, 1, 0 };
	static int[] dc = { 0, 1, 0, -1 };

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(br.readLine());

		map = new int[N][N];

		for (int i = 0; i < N; i++) {
			char[] cst = br.readLine().toCharArray();
			for (int j = 0; j < N; j++) {
				map[i][j] = cst[j] - '0';
			}
		}

		int cnt = 0;
		int[] cntList = new int[N*N];
//		List<Integer> cntList = new ArrayList<Integer>();
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < N; j++) {
				if (map[i][j] == 1) {
					cnt++;
					cntList[cnt-1] = bfs(i, j, cnt + 1);
//					cntList.add(bfs(i, j, cnt + 1));
				}				
			}
		}
		
		cntList = Arrays.copyOf(cntList, cnt);
		Arrays.sort(cntList);
//		Collections.sort(cntList);
//		cntList.sort((num1, num2) -> Integer.compare(num1, num2));
		
		StringBuilder sb = new StringBuilder();
		System.out.println(cnt);
		for (int i = 0; i < cnt; i++) {
			sb.append(cntList[i]).append("\n");
//			System.out.println(cntList[i]);
		}
		System.out.println(sb);
	}

	public static int bfs(int cr, int cc, int g) {
		Queue<int[]> que = new LinkedList<int[]>();
		que.offer(new int[] { cr, cc });
		map[cr][cc] = g;
		int result = 1;

		while (!que.isEmpty()) {
			int[] cur = que.poll();
			int r = cur[0];
			int c = cur[1];
			for (int d = 0; d < 4; d++) {
				int nr = r + dr[d];
				int nc = c + dc[d];
				if (!check(nr, nc))
					continue;
				if (map[nr][nc] == 1) {
					que.offer(new int[] { nr, nc });
					map[nr][nc] = g;
					result++;
				}
			}
		}
		return result;
	}

	private static boolean check(int r, int c) {
		return r >= 0 && r < N && c >= 0 && c < N;
	}

}

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
# 아이디어
# bfs
# for문 돌면서 집인데 visited안간거 찾아서 bfs돌림
# bfs함수 리턴값 : 같은 구역에 있는 집 수

from collections import deque

dr = [1, -1, 0, 0]
dc = [0, 0, 1, -1]

def bfs(fr, fc):
    # 초기화
    result = 1
    queue = deque([[fr, fc]])
    visited[fr][fc] = 1

    # queue빼기
    while queue:
        q = queue.pop()
        cr, cc = q[0], q[1]
        for d in range(4):
            nr, nc = cr + dr[d], cc + dc[d]
            if 0<=nr<N and 0<=nc<N:
                if visited[nr][nc]: # 이미 방문했으면
                    continue
                if not arr[nr][nc]: # 집이 아니면
                    continue
                queue.append([nr, nc])
                visited[nr][nc] = 1
                result += 1
    return result




N = int(input())
arr = [list(map(int, input())) for _ in range(N)]
visited = [[0] * N for _ in range(N)]
ans = []

# 입력 끝! 로직 시작!

for r in range(N):
    for c in range(N):
        if arr[r][c] and not visited[r][c]:
            ans.append(bfs(r, c))

ans.sort()
print(len(ans))
print(*ans, sep="\n")
